### DISCUSSION TITLE
Discussion about open source licenses/Chromium forks

### DISCUSSION THREAD

**Discussion by uazo** [bromite/bromite#2144] [General]
@csagan5  allow me to communicate my position, I could not do it before simply for lack of tools, I will not write more on the subject.

I align myself with your position for what the objectives are: I am also interested only in being able to contribute technically to the creation of a better browser and sleep peacefully, I'm not going to participate in anything else.

For what it can serve, you have all my support for the choice of giving a first chance to anyone to be able to collaborate.
I consider it a sensible choice and in line with the arguments of the open source.
I also think no one should feel they have the right to impose otherwise, especially in the violent ways I have been able to read: so if the choice (which I agree with) will bring us consequences in the development of bromite, we can live with that.

Going beyond the events, if the problem is the use of my patches, I consider myself a person with whom it is possible to discuss: if exists the possibility of a compromise that allows any project to exploit my patches guaranteeing me (and you) not to have any kind of problem, than the choice of gpl or mit or any other license is not the problem, please @thestinger let me know.
to date I rely on the indications that I have understood over time: gpl3 seems to relieve me more from potential judicial problems and
allows to keep my work open to anyone, whatever changes are subsequently made to the base of my patch.

If, on the other hand, the real problem is another, as far as I'm concerned, I stay away from it: if then, even for no reason, I am not welcome in the various communities, perhaps I am of little interest, or perhaps I am wrong, time will tell. honestly, I don't think it would change anything in the way I contribute to open source for now.

in any case, far be it from me the intention of creating rifts in the world of open source that I would have said was driven by the mere desire to cooperation.

**Comment by csagan5** [comment]
> @csagan5 allow me to communicate my position, I could not do it before simply for lack of tools, I will not write more on the subject.

I have to lock all the conversations about this, they are off-topic, sorry. Even with this new thread you are fuelling a sort of "flamewar" which benefits no one, and more discussions may spawn, which I will have to lock anyways. This drains energy/attention from working on the project (for all of us).

> For what it can serve, you have all my support for the choice of giving a first chance to anyone to be able to collaborate.
> I consider it a sensible choice and in line with the arguments of the open source.

Thank you; let me also specify (to avoid someone misquoting you, since as a major Bromite contributor you have been quoted somehow as representing it officially - which is not the case) that I do not ask for support and do not need it: in the worst case that you (or anyone else) decides to not contribute to this project or fork it, exactly nothing will happen: that is why open source exists! Debian did not "declare war" to Canonical when they started using their packages 17 years ago. Of course I welcome all contributions because without that open source slowly dies...that is another reason why I want to facilitate ASAP that everyone (or at least as many people as possible) can make their own builds.

> I also think no one should feel they have the right to impose otherwise, especially in the violent ways I have been able to read: so if the choice (which I agree with) will bring us consequences in the development of bromite, we can live with that.

I wouldn't say "violent ways", violence in words (allusion) perhaps. There are no consequences for Bromite except for the reduced "brain power" if less people want to come over here in this project and report bugs, suggest improvements, discuss technical problems. This place remain open to that* as it has always been.

*= it has never been to open to quickly submitted issues, submitting an issue requires some work, and that will not change

> Going beyond the events, if the problem is the use of my patches, I consider myself a person with whom it is possible to discuss: if exists the possibility of a compromise that allows any project to exploit my patches guaranteeing me (and you) not to have any kind of problem

All patches submitted here are under GPLv3; you can re-license to **other projects** the patches that you are author of under any other license. It has always been like that. Thinking that there can be any problem is a net effect of the confusion created by all the BS mentioned in #2102 and related discussions, please don't fall for that.
How does re-licensing works? You simply pull a patch from your own personal copy (from where you originally created it, not from another repository) and stick to it the license that you want. You are technically not re-licensing a GPLv3 patch as BSD/MIT etc (which you cannot) but creating a new patch from your original "pre-licensed" work which is present on your devices.

> to date I rely on the indications that I have understood over time: gpl3 seems to relieve me more from potential judicial problems and
allows to keep my work open to anyone, whatever changes are subsequently made to the base of my patch.

The only situation to avoid is patches without a license because someone could claim them as their (proprietary) work with some dirty tricks; you should read more on the topic and not rely on my word or anyone else's.

> If, on the other hand, the real problem is another, as far as I'm concerned, I stay away from it: if then, even for no reason, I am not welcome in the various communities, perhaps I am of little interest, or perhaps I am wrong, time will tell. honestly, I don't think it would change anything in the way I contribute to open source for now.

Vanadium will probably welcome your patches if you re-license them as anything non-GPL, but it is sad that you would decide to change the license after all of this, because it would look like "scare tactics" working on you. And if you think of it, GPL was created **exactly to avoid this** (quoting from [this comment](https://github.com/bromite/bromite/pull/2102#issuecomment-1155832810)):

> The license for Vanadium can always be changed to have stricter requirements going forward.

Having said that: the Bromite project does not accept/deny patches based on which other projects you contribute to, so you are absolutely free to do that and I would not even have any personal grudge about that (because I think that [open source does not work any differently](https://github.com/bromite/bromite/issues/2141#issuecomment-1159398706)).

> in any case, far be it from me the intention of creating rifts in the world of open source that I would have said was driven by the mere desire to cooperation.

Sorry, you have not created rifts in the open source world :slightly_smiling_face: the scale here is much, much smaller; and everyone can use which license they like on their work. Once submitted here it will however be licensed under GPLv3, as it has always been. If that scares away some contributions then I am sorry for that, but if you can read all the facts this whole incident is perfect evidence of the importance of the GPL license.

**Reply by uazo** [reply]
> Even with this new thread you are fuelling a sort of "flamewar" which benefits no one, and more discussions may spawn, which I will have to lock anyways.

I'm sorry, but I wanted to let you know what I think. lock this discussion now as well

> I do not ask for support and do not need it

I realize this

> I wouldn't say "violent ways", violence in words (allusion) perhaps.

sure.

> Vanadium will probably welcome your patches if you re-license them as anything non-GPL,

vanadium changed the license to gpl2 just yesterday, I'll try to understand what that change means

**Reply by csagan5** [reply]
> vanadium changed the license to gpl2 just yesterday, I'll try to understand what that change means

https://github.com/GrapheneOS/Vanadium/commit/773b541835b22212d5462ff75ee457be0fbcf745
Was not aware; it means that any new patch that is contributed there is GPLv2, the old patches still have the old license.

The two licenses are not compatible: https://opensource.stackexchange.com/questions/1777/why-is-gplv2-incompatible-with-gplv3

However the Chromium project itself already contains GPLv2 and GPLv3 (and another bunch of licenses, see the old list here: https://web.archive.org/web/20081026150300/http://code.google.com/chromium/terms.html): https://source.chromium.org/search?q=%22GPL%20version%203%22&sq=&ss=chromium%2Fchromium%2Fsrc

We will think of any possible problem when it is the case that someone submits a patch saying that license is GPLv2.

For your own work you can always re-license it with any license, as explained before.

**Reply by uazo** [reply]
but allow me the stupid question.

does the `.patch file` have a license or the final code produced by the patch that acquires it?

because, considering that I am interested in that the patch code (the .patch file) remains free and open, I do not care if the product of the patch is closed and not modifiable, it is chromium after all. mi spiego bene?

**Reply by csagan5** [reply]
It is not a stupid question, it is the core question (that is why I have not locked the discussion as well). The patch files are licensed, or better: the code within those patch files.

> I do not care if the product of the patch is closed and not modifiable, it is chromium after all. mi spiego bene?

I do care, but the Chromium APK (which is the binary) cannot be considered a GPLv2 or GPLv3 binary (since it contains software of so many different licenses). So the application of the binary clauses is probably effective only if you build some other software that uses the patches.

**Reply by uazo** [reply]
> I do care, but the Chromium APK (which is the binary) cannot be considered a GPLv2 or GPLv3 binary (since it contains software of so many different licenses). So the application of the binary clauses is probably effective only if you build some other software that uses the patches.

but all the patches act on the same code, which is chromium, which is open and freely usable by everyone.
and the patch alone is useless.
so what are we talking about? the binary product is always distributable regardless of which patches you have applied, because it is google licensed, correct? so if the patch is gpl2 or 3 or any other, it doesn't matter as if the sources remain open, what am I wrong?

**Reply by csagan5** [reply]
> but all the patches act on the same code, which is chromium, which is open and freely usable by everyone.

Open and freely usable second its default license that you can find here: https://source.chromium.org/chromium/chromium/src/+/main:LICENSE

It is not public domain.

> so what are we talking about? the binary product is always distributable regardless of which patches you have applied, because it is google licensed, correct?

No, it is not licensed by Google, Chromium's copyright is (by default) by "The Chromium Authors", so it is licensed by them directly. Yes the binary can always be distributed but you cannot apply both GPLv2 and GPLv3, that is what is mentioned here: https://opensource.stackexchange.com/questions/1777/why-is-gplv2-incompatible-with-gplv3 e.g. you cannot go to someone using the Chromium source in a fork and enforce both licenses.

> so if the patch is gpl2 or 3 or any other, it doesn't matter as if the sources remain open, what am I wrong?

You need to do some reading on the topic, I am not the right person to explain this but will try; when you write code the copyright is yours, nobody can take away that creation from you (unless you work for another company, then your NDA might say that they own your creations).
When you license it then you make a copy of your work and attach a license to it. Note that this copy does not affect your original creation but the copy you are providing to others. By making this you do not cease to be the creator of the original and so you can still license other copies.

Now, if license is BSD/MIT/etc anyone receiving the code and the license can do what the license says; then they can make modifications to your code and never release them (making them proprietary) and/or sell some product that contains it (there is nothing ethically wrong with this, although most people ignore this aspect when picking BSD/MIT and similar licenses).

One fundamental difference with GPL (there are other differences) is that instead GPL says that anyone releasing a binary which was built using your GPL-protected work should - if someone asks - provide the source code used, including the modifications. They cannot refuse this. GPLv3 adds even other clauses.

So if your concern is that you somehow would "lose" the right to use your own work: that can only happen if you do not attach any license (public domain) and someone tries to pretend that they made that exact same work before you (I do not know exactly how this would play out but it is possible and has happened, again I suggest to do research about it). It cannot happen if you have used an approved license from here: https://opensource.org/licenses

Bromite uses [GPL](https://opensource.org/licenses/gpl-license) for the extra clauses. But do not confuse "GPL" with "open source licenses". GPL is just one of many licenses, see above links.

So you can always dual-license your work in as many licenses as you like, but the opposite is not true: if anyone receives code covered by GPL they should not re-license it as BSD/MIT or any other one, unless original author allows it. This quickly becomes a problem when it is not easy/possible to reach all the authors so big projects usually ask contributors to "give up" their intellectual copyright rights to the the project so that they're free to re-license it later on. This is not the case for Bromite contributions.

**Reply by csagan5** [reply]
You can find the CLA required to contribute to Chromium here: https://cla.developers.google.com/about/google-individual?csw=1

After you accept this CLA then if your contributions are accepted upstream they end up with [this 3-clause BSD license](https://opensource.org/licenses/BSD-3-Clause)

(But since you accepted the CLA, at any moment in time in future your work can be re-licensed with any other license, "bypassing" the licensing mechanisms explained before)

**Reply by uazo** [reply]
> One fundamental difference with GPL (there are other differences) is that instead GPL says that anyone releasing a binary which was built using your GPL-protected work should - if someone asks - provide the source code used, including the modifications. They cannot refuse this. GPLv3 adds even other clauses.

ok, I understood that. 
however not even bromite is formally gpl3, since the patch that transforms chromium into bromite is missing, is not public.
and in any case by applying the digital signature I am not allowed to modify the binary, which I understand that the gpl should allow.
so, what changes between choosing a v2 or a v3 against vanadium, or any other gpl project?

> So if your concern is that you somehow would "lose" the right to use your own work

no, the first concern is any improvements to my patch not being made public

**Reply by csagan5** [reply]
> > One fundamental difference with GPL (there are other differences) is that instead GPL says that anyone releasing a binary which was built using your GPL-protected work should - if someone asks - provide the source code used, including the modifications. They cannot refuse this. GPLv3 adds even other clauses.
> > 
> ok, I understood that.
> however not even bromite is formally gpl3, since the patch that transforms chromium into bromite is missing, is not public.

All patches combined transform Chromium into Bromite, except for the branding parts (icon) - if you are referring to that. I do not publish it to make it slightly more difficult to create clones branded as Bromite but this does not change anything regarding the licensing status of Bromite: all the published patches (and the source code they contain) are under GPLv3 and the binary is instead the result of a mix, same as Chromium.

> > So if your concern is that you somehow would "lose" the right to use your own work
> 
> no, the first concern is any improvements to my patch not being made public

Yes, then GPL protects against that. You can release work in GPLv2 with the upgrade clause, then it is compatible with both versions. See this explanation here: https://opensource.stackexchange.com/a/6263

The license used here https://github.com/GrapheneOS/Vanadium/commit/773b541835b22212d5462ff75ee457be0fbcf745 has no such clause but your patch would be compatible.

> and in any case by applying the digital signature I am not allowed to modify the binary, which I understand that the gpl should allow.

No, the GPL does not say anything about modifying binaries. It says that if you receive a binary produced using GPL sources you can ask for the sources. GPLv3 has extra clauses against "hardware locking" of devices.

> so, what changes between choosing a v2 or a v3 against vanadium, or any other gpl project?

I do not think the extra clauses of GPLv3 currently apply for Chromium forks, but I am not a lawyer; I picked GPLv3 back then because of what is explained here: https://www.gnu.org/licenses/quick-guide-gplv3.en.html

If in future Android (or any other fork, like GrapheneOS etc) would forbid the Bromite APKs to be installed there but at the same time they ship software using Bromite code, then you could say that they are in breach of the GPLv3 clause against tivoization, it's explained all there in the link above. That's it for the differences between GPLv2 and GPLv3. I do not think that Android/GrapheneOS/other forks currently have such restrictions.

**Reply by uazo** [reply]
> Yes, then GPL protects against that. You can release work in GPLv2 with the upgrade clause, then it is compatible with both versions.

perfect i like it, i think my patches will be "gpl2 or later", it means trusting version 4 when it will be there, but that's okay for me.
this would also allow vanadium not to ask me for permission to use it, "manco fossi bill gates :)"
we should figure out if you prefer it to have a separate repo (which I wish I didn't have) or if we can put the information directly into the patch, as you suggested.

**Reply by csagan5** [reply]
> perfect i like it

I do not because I suspect this license was chosen on purpose to try to not be compatible with Bromite's GPLv3, so that the patches sharing goes one way only. But it's just speculation.

> we should figure out if you prefer it to have a separate repo (which I wish I didn't have) or if we can put the information directly into the patch, as you suggested.

All patches here are GPLv3; you will have to put them in a repository with license "GPLv2 or later". I am already planning to put in each patch something like `License: GPLv3` and `Original license: ...`, but that is the best I can do. Not planning to change the license of Bromite.

**Reply by michitux** [reply]
> I do care, but the Chromium APK (which is the binary) cannot be considered a GPLv2 or GPLv3 binary (since it contains software of so many different licenses). 

According to my understanding (I'm not a lawyer, this isn't legal advice) that matches what's written [here](https://opensource.stackexchange.com/questions/6283/how-can-a-gplv2-dependency-affect-licensing-of-adjacent-components-that-form-an), as soon as you're using GPL code, you must consider your binary as GPL (or another license that fulfills all conditions listed in the GPL) even if parts of your code were under a more permissive license. All licenses must be [compatible with the GPL](https://law.stackexchange.com/a/19167), otherwise you have a problem. As soon as you include GPLv2-only code, your binary cannot also include GPLv3-only code and vice-versa, as the two licenses are incompatible. If you include both, the binary cannot be legally distributed anymore. Or in other words, the license issue exists primarily when you distribute binaries, not when you just distribute a collection of patches.

**Reply by csagan5** [reply]
> All licenses must be [compatible with the GPL](https://law.stackexchange.com/a/19167), otherwise you have a problem. As soon as you include GPLv2-only code, your binary cannot also include GPLv3-only code and vice-versa, as the two licenses are incompatible. If you include both, the binary cannot be legally distributed anymore.

A look at `chrome://credits` should be sufficient to see how this is done for Chromium; it should not be any different for Bromite. All the components there seem to be in a permissive license or multi-licensed.

**Reply by uazo** [reply]
> I do not because I suspect this license was chosen on purpose to try to not be compatible with Bromite's GPLv3, so that the patches sharing goes one way only. But it's just speculation.

unfortunately it is a legitimate doubt. it looks like a "guerra tra poveri", if I can help tone it down, that's fine by me

> Not planning to change the license of Bromite.

absolutely, never asked. 

> you will have to put them in a repository with license "GPLv2 or later"

okay, I will prepare an ad hoc repo for my patches, maybe we agree on the list first because maybe you have given me some by mistake (like the always incognito)

**Reply by csagan5** [reply]
> you will have to put them in a repository with license "GPLv2 or later"
> 
> okay, I will prepare an ad hoc repo for my patches, maybe we agree on the list first because maybe you have given me some by mistake (like the always incognito)

In practice if you do not want people to have to ask you every time the repository is the most practical solution, because anyone can just take it from there with the correct license. In Bromite I was thinking to have in the patch file description for your patches "Original license: GNU GPLv2 or later" and "License: GNU GPL v3", just to avoid to formally introduce multi-licensing, but the difference is very subtle. I think you can also avoid the separate repository and whoever wants to use the patch should be able to understand the licensing this way. Only problem would be if the version of Bromite is very different than your version...and you would have to remember to not copy from Bromite into your repository in case it was someone else updating the patch (substantially).

As you can see all of this is an immense waste of time (because it will change nothing in practice and not benefit any person/ego) and it has already taken away time from the actual development; so feel free to proceed as you like, I don't mind either ways and I am sorry for the extra work/thinking this has caused so far.

**Reply by uazo** [reply]
> In Bromite I was thinking to have in the patch file description for your patches "Original license: GNU GPLv2 or later" and "License: GNU GPL v3", just to avoid to formally introduce multi-licensing, 

I agree.

> I think you can also avoid the separate repository and whoever wants to use the patch should be able to understand the licensing this way

thank you, for me it would be less work.

> As you can see all of this is an immense waste of time (because it will change nothing in practice and not benefit any person/ego) and it has already taken away time from the actual development; so feel free to proceed as you like, I don't mind either ways and I am sorry for the extra work/thinking this has caused so far.

but do not worry! it's not wasted time, I learned one more thing.

> After you accept this CLA then if your contributions are accepted upstream they end up with [this 3-clause BSD license](https://opensource.org/licenses/BSD-3-Clause)

instead, let's go ahead, you told me about the CLA, which I later read ... I don't like signing anything, especially if it means giving superpowers to google, but I'm sorry I can't contribute to chromium.
i tried asking this https://bugs.chromium.org/p/chromium/issues/detail?id=1284380#c10
